static/app/views/dashboardsV2/detail.tsx

```diff
@@ -254,8 +254,48 @@ class DashboardDetail extends Component<Props, State> {
   };
 
   onAddWidget = () => {
-    const {organization, dashboard} = this.props;
-    openDashboardWidgetLibraryModal({organization, dashboard});
+    const {organization, dashboard, api, reloadData, location} = this.props;
+    this.setState({
+      modifiedDashboard: cloneDashboard(dashboard),
+    });
+    openDashboardWidgetLibraryModal({
+      organization,
+      dashboard,
+      onAddWidget: (widgets: Widget[]) => {
+        const modifiedDashboard = {
+          ...this.state.modifiedDashboard!,
+          widgets,
+        };
+        this.setState((state: State) => ({
+          ...state,
+          widgetToBeUpdated: undefined,
+          modifiedDashboard,
+        }));
+        updateDashboard(api, organization.slug, modifiedDashboard).then(
+          (newDashboard: DashboardDetails) => {
+            addSuccessMessage(t('Dashboard updated'));
+            this.setState({
+              dashboardState: DashboardState.VIEW,
+              modifiedDashboard: null,
+            });
+
+            if (reloadData) {
+              reloadData();
+            }
+            if (dashboard && newDashboard.id !== dashboard.id) {
+              browserHistory.replace({
+                pathname: `/organizations/${organization.slug}/dashboard/${newDashboard.id}/`,
+                query: {
+                  ...location.query,
+                },
+              });
+              return;
+            }
+          },
+          () => undefined
+        );
+      },
+    });
```

Comment:
nit: The code here works but I suspect that you might be able to remove all the `setState` lines from here since the state attributes are being used to track pending changes in the Edit view and the `onAddWidget` event is exclusive from the Edit View. My guess is that these state attributes probably won't be used for the custom widget tab either. It could reduce the length of this function a bit. Let me know if I've missed something!

```suggestion
    openDashboardWidgetLibraryModal({
      organization,
      dashboard,
      onAddWidget: (widgets: Widget[]) => {
        const modifiedDashboard = {
          ...cloneDashboard(dashboard),
          widgets,
        };
        updateDashboard(api, organization.slug, modifiedDashboard).then(
          (newDashboard: DashboardDetails) => {
            addSuccessMessage(t('Dashboard updated'));

            if (reloadData) {
              reloadData();
            }
            if (dashboard && newDashboard.id !== dashboard.id) {
              browserHistory.replace({
                pathname: `/organizations/${organization.slug}/dashboard/${newDashboard.id}/`,
                query: {
                  ...location.query,
                },
              });
              return;
            }
          },
          () => undefined
        );
      },
    });
```

Reply:
sorry for the lengthy diff lol